### ISSUE TITLE
# Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr ************ in tid **** (SGen worker)

### ISSUE THREAD

**Issue by dcrespo-isaac** [dotnet/runtime#105389]
### Android framework version

net8.0-android

### Affected platform version

.NET 8.0.100, .NET 8.0.303

### Description

Our Android application started to crash randomly 3 months ago when we upgraded it to .net6. 
We tried to add more logging, analyze the logcat, reproduce... but all in vain. 

In an attempt to stabilize the app, we have upgraded to .net8.0-android and the number of crash was significantly reduced.
This problem happens on different kind of tablets, and we excluded the idea that it could come from firmware or hardware 
because they did not change and with .net framework (Xamarin) everything was running smoothly (1 crash rarely vs hundreds per day now).

Our dev and testing team (qa) are not able to reproduce the problem. Looking at the logs ([SIGSEGV_Samples_logcat-with-symbols.txt](https://github.com/user-attachments/files/16354417/SIGSEGV_Samples_logcat-with-symbols.txt)), we see that all of the SIGSEGV (SEGV_MAPERR) happens on the SGen worker thread but give different backtraces.
At the end of the log file, you’ll find a SIGABRT from SGen. We’ve kept it there for reference, in case it’s related to the SIGSEGV issues.

We used `dotnet-symbol` to retrieve a debug version of libmonosgen-2-0.so (with symbols). Then we run `ndk-stack` on our
logcat to symbolize the backtraces.

Application catches nothing. So we don't have any other traces to give from it. The logcat contains the most detailled 
ones we have.

FYI, we did split our crashes into 3 categories, we're not sure that they related to the SIGSEGV... 
but to give you all data, the other 2 are : 
- Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid ***** (|ANR-WatchDog|), pid ***** (com.isaac.onthego)
- Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid ***** (enableInternal), pid ***** (com.android.nfc)

What we tried : 

- Reproducing by running our acceptance tests on the tablet
-> Nothing happens unless we run the day long tests and it's very rare. It does not give more information than what we have in production.
- Upgrading .net version (from 8.0.100 to 8.0.303)
-> Nothing changed.

- **Revert to .Net Framework (Xamarin)**
-> We did it in order to identify if it's related to our changes or the framework. **And it worked !!!  
The number of crash were back to normal (before .net6 and .net8).** 

What we are still probing :  
 
- Disabling concurrency on Mono Garbage collector and change the bridge-implementation to old  
-> We’ve seen numerous issues and discussions about the stability of SGen. We are hoping that this change can make a difference, as the backtraces seem to point to a problem that could be within it. If this works, it could allow us to stay with .net8 until a fix is found.

Thank you in advance for your help. We will remain vigilant if you need more details.

### Steps to Reproduce

Unfortunetaly we are unable to reproduce the problem in step by step as it appears randomly on production.

### Did you find any workaround?

Currently, reverting to the .Net Framework appears to be effective for us.

### Relevant log output

```shell
See attached files of the description
```

**Comment by akoeplinger** [comment]
@BrzVlad looks like a GC issue

**Comment by grendello** [comment]
@dcrespo-isaac Scan the logs for messages containing strings similar to `* Assertion at` which are tagged either with an empty string, `DOTNET` or a fragment of your application's ID.  The message will also contain a file path which should have `mono` component in it.

**Comment by dcrespo-isaac** [comment]
Thank you for the triage. 

I've found the logcat I had analyzed for the SIGSEGV errors. In the [logcat-Assertion.txt](https://github.com/user-attachments/files/16366700/logcat-Assertion.txt), you'll find two samples of SIGABRT preceded by :

```Assertion at /__w/1/s/src/mono/mono/metadata/object.c:4410, condition `is_ok (error)' not met, function:mono_unhandled_exception_internal, (null) assembly:System.Private.CoreLib.dll type:ArgumentNullException member:(null)```

Additional notes:

- The SIGABRT assertion error is always the same, but it sometimes occurs on the `ANR_WatchDog` thread and other times on what appears to be the main thread `com.isaac.onthego`.
- The two samples show errors just after the start of the application. It’s true that some errors occur then, but others occur at different moments – sadly at random times – throughout the lifetime of the application.
- The number of SIGSEGV crashes is substantially higher than that of SIGABRT.
-> We attempted to identify a pattern between SIGSEGV and SIGABRT, but found nothing significant. This means that they don’t precede or follow each other in any particular way.

I will continue to update the issue if I encounter any other assertion errors.

**Comment by lambdageek** [comment]
@dcrespo-isaac I don't have any insight for the SIGSEGV, but for SIGABRT - does your app have a `AppDomain.CurrentDomain.UnhandledException` event handler installed?  the assertion on object.c:4410 looks like it's due to an unhandled null reference exception during the execution of the unhandled exception handler.

**Comment by dcrespo-isaac** [comment]
Hi @lambdageek,

First and foremost, thank you for your response. It is greatly appreciated. We have identified the origin of the SIGABRT:

- There was indeed a null reference that was triggered due to a race condition at startup.
- We discovered a setter in ANR_WatchDog that allows us to handle an ANR in our own way, rather than allowing the tool to crash our application.

Regarding this open issue, we continue to experience problems with SIGSEGV, which is the primary subject of this issue and, regrettably, the cause of the majority of our crashes.

We are currently deploying our reverted version to Xamarin, as it has proven to be significantly more stable. We will remain vigilant for any responses, fixes, or requests for additional information regarding this issue.

**Comment by dcrespo-isaac** [comment]
Hi Team, @mangod9 ,

We noticed a few weeks ago that the issue was moved to the “Future” milestones.

While “No due date” is quite clear, it leaves us with a significant span of uncertainty, especially since Xamarin is no longer supported. We would like to manage our expectations regarding this issue. Will it be fixed or at least investigated by your team of experts?

Currently, we have had to completely rollback all our projects to the old .NET Framework/Xamarin to reduce crashes due to SIGSEGV.

Please let us know if there are any troubleshooting techniques or tools we can use to provide you with more information, or ideally, find a fix ourselves.

We look forward to hearing more on this thread,
Best regards to your team.

**Comment by mangod9** [comment]
Hello @dcrespo-isaac, you mention that the issue is reduced in 8, but still occurs occasionally?

**Comment by dcrespo-isaac** [comment]
Hi @mangod9,

It appears there may have been a misunderstanding regarding the issue.é
The concern is about an **increase** in the number of SIGSEGV occurrences on .NET 8 (specifically on the SGen Worker), not a decrease.

Our workaround was to revert to Xamarin because it:

- Fixed the SIGSEGV errors on the SGen Worker, which is the primary focus of this thread.
- Generally reduced the number of crashes and SIGSEGVs (null dereference). These errors are challenging to troubleshoot, but we believe they are not an issue with .NET. Consequently, we did not open an issue for them.

**Comment by mangod9** [comment]
Hi @vitek-karas, since this is on Android can you please provide guidance on diagnostic steps. Is there a way to collect some dumps when the failure occurs?

**Comment by BrzVlad** [comment]
This issue looks like duplicate of https://github.com/dotnet/runtime/issues/100311. Did disabling concurrent GC lead to fixing of the crashes ? This issue seems somewhat impossible to investigate without a repro. However, I think we can give an attempt to investigate whether something is wrong with the GC by trying some debug flags like `MONO_GC_DEBUG=check-remset-consistency,mod-union-consistency-check,collect-before-allocs,verify-before-allocs`. Some of this options might be bitrotten so it would be ideal if we could have access to run the application ourseleves.

**Comment by dcrespo-isaac** [comment]
Hi @BrzVlad,

I reviewed issue #100311, which pertains to SIGABRT. We no longer face issues with SIGABRT; this thread is actually about SIGSEGV. 
However, @pijnappel has created an issue that seems similar to ours: dotnet/maui#21479. 
Additionally, dotnet/android#9402 appears to be related as well.

We plan to beta test the version without concurrent GC in a few days and will keep everyone updated. Thank you for the MONO_GC_DEBUG options; I was unaware of some of them, and they will be helpful.

What exactly would you need to run everything on your end?

**Comment by grendello** [comment]
@BrzVlad Replacing a shared library in the apk requires that you re-align and re-sign the archive, or it won't install and run on the device/emulator.  Signing process records sha256 hashes of the shared libraries.  When you need to do that, ping us on Discord in the #android channel, we'll help you out :)

**Comment by BrzVlad** [comment]
@dcrespo-isaac In addition to testing with concurrent gc disabled, I think it would be worthwhile to keep concurrent gc enabled but use the flags `MONO_GC_PARAMS=evacuation-threshold=0,no-precleaning` which disables some functionality that seems involved in the crash reports.

**Comment by dcrespo-isaac** [comment]
@BrzVlad, 

Thank you. We will develop a plan to test the different options and will get back to you with the data. At the very least, we will have two versions:

- Concurrent GC disabled, debug improved with `check-remset-consistency,mod-union-consistency-check,collect-before-allocs,verify-before-allocs`
- Concurrent GC enabled, debug improved with `check-remset-consistency,mod-union-consistency-check,collect-before-allocs,verify-before-allocs`, options `evacuation-threshold=0,no-precleaning`

Additionally, I am compiling information to create a brief documentation on how you can test our .apk on your end.

**Comment by BrzVlad** [comment]
Don't use the `mod-union-consistency-check`, there seems to be some issues with it. Also note that `collect-before-allocs,verify-before-allocs` are absurdly heavy, maybe use them just for a few runs manually to see if they happen to report anything, I don't think it would be worthwhile to stress test with it. I would say it is best to first run without any debug flags just to see whether the gc options might fix the crash and only afterwards attempt to run with gc flags on the crashing configurations.

**Comment by dcrespo-isaac** [comment]
Hi, Quick update,

We're sorry for the delay, we should be able to test and give you more data by next week.

**Comment by BrzVlad** [comment]
My current understanding is that concurrent GC was enabled by default with the new .NET versions (5+). Concurrent GC was supported in legacy Xamarin but you needed to explicitly add it in the csproj (or click the option in the project editor). This would mean that using this configuration for .NET is not really a change in behavior and you shouldn't encounter any problems with pause times.

The `check-remset-consistency` option had some issues which I addressed recently. The new dotnet version with the fix should be available in 1-2 months. I would recommend its usage only once the new version is available, but for the sole purpose of helping us investigate the issue with concurrent GC that nobody is able to reliably reproduce so far. This debug option can end up more aggressively killing the runtime if a problem is found in the GC, so I would understand if you wouldn't want to deploy an app with this enabled. However, giving it a try for localized testing would be appreciated.

**Comment by IainS1986** [comment]
@BrzVlad How are you building with the concurrent GC disabled?

Are you just doing it in the csproj configuration with


    <AndroidEnableSGenConcurrent>false</AndroidEnableSGenConcurrent>

Or is there other build params/steps?

**Comment by Th3L0x** [comment]
I'm encountering an issue that seems specific to Xiaomi devices. The error details are as follows:

```
[My.App] * Assertion at /__w/1/s/src/mono/mono/metadata/sgen-tarjan-bridge.c:1176, condition `xref_count == xref_index` not met, function: processing_build_callback_data, xref_count is 642 but we added 638 xrefs  
[monodroid] * Assertion at /__w/1/s/src/mono/mono/metadata/sgen-tarjan-bridge.c:1176, condition `xref_count == xref_index` not met, function: processing_build_callback_data, xref_count is 642 but we added 638 xrefs  
[monodroid] Abort at mono-log-adapter.cc:46:3 ('static void xamarin::android::internal::MonodroidRuntime::mono_log_handler(const char *, const char *, const char *, mono_bool, void *)')  
[libc] Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 20651 (Agitur.DeepDate), pid 20651 (My.App)  
```

This error does not occur on other Android devices or on iOS, only on Xiaomi devices. 

This config: `<AndroidEnableSGenConcurrent>false</AndroidEnableSGenConcurrent>` didnt solved it.

**Comment by BrzVlad** [comment]
@IainS1986 That is correct.

@Th3L0x That is a completely unrelated issue. See https://github.com/dotnet/runtime/issues/106410 for workaround until we are able to investigate it.

**Comment by Syed-RI** [comment]
Would this comment be worth investigating? https://github.com/dotnet/runtime/issues/106410#issuecomment-2289648143

**Comment by byteriderhun** [comment]
Hi!
I also have problem with this Fatal signal 11 (SIGSEGV) error.
My android app produces it very easy at a certain point. 
If you would like to experience it, I can add you to the internal tester list and you can download it from play store.

**Comment by Syed-RI** [comment]
> Hi! I also have problem with this Fatal signal 11 (SIGSEGV) error. My android app produces it very easy at a certain point. If you would like to experience it, I can add you to the internal tester list and you can download it from play store.

Hey do you have big (in Mbs) network/IO operation in the app? Please feel free to add me to test.

**Comment by byteriderhun** [comment]
Yes I have. I need your email, than you can try.

**Comment by Syed-RI** [comment]
> Yes I have. I need your email, than you can try.

please add generic.test711@gmail.com :)

**Comment by byteriderhun** [comment]
[https://play.google.com/apps/internaltest/4701715275062451291](url)

You can download the app from here. I hope it will run on your device.
It is a hungarian app, but I write you what to do.
When it started and you can see the icons, from the right you can pull in login window.
Press "Bejel" button to log in. 
Use "Felhasználó név:" test  "Jelszó" test
Press the "Restaurant Sales" icon. Select "Ok"
On the bottom you will see "Betöltés folyamatban". This means loading. When it dissapears select
"Számla". 
You will see several rows. 
In the "Asztalok, Személyek" select a yellow item. For example "32". It has a lot of items.
Select row "Tartalom". This is the content of the table.
Now here we are! Clicking on items they will go to the next row called "Külön fizetendő"
If you select "Külön fizetendő" and click on item it will go back.
If you do this fast it will randomly crash with this [Fatal signal 11 (SIGSEGV)] error.
In previous versions it crashed more frequently, but I removed network calls to eliminate the problem.
Now crash happens not so fast but if you randomly do the clicking on items it will happen in some minute.
In the last 5 minutes I could crash it 3 times.

**Comment by feric-isc** [comment]
Hello all. 
@dcrespo-isaac is away for a while and I will be taking over on our side for a while. 

Are there any other news related to this issue? 

On our side, we deployed the .NET8 version with concurrent gc disabled and it did allow us to get back to "xamarin-like" levels of the SIGSEGV/SIGQUIT/SIGABRT levels. So the good news is we were able to jump on the .NET side of things.

The issue is, I still find that the level of these SIGSEGV/SIGQUIT/SIGABRT to be unreasonably high. 1 in 5 of our devices will get this exception every day (unfortunately we're not able to pinpoint when...very random).

Browsing through this site, I found a few other issues like https://github.com/dotnet/runtime/issues/100311 or https://github.com/dotnet/android/issues/8253#issuecomment-2380714158. Could these be related and if so, do you have any other suggestion on how to reduce these exceptions?

I also saw this issue: https://github.com/dotnet/android/issues/9365. We do have some ANRs that seem to have increased in this new version... I will be following this thread as well as I see it is very active but I'm guessing they couldn't have the same root cause?

**Comment by BrzVlad** [comment]
@feric-isc What are the stacktraces for the current crashes that you are observing ?

**Comment by feric-isc** [comment]
Hi @BrzVlad,

We track our errors through Crashlytics. 

For SIGSEGV errors, here are a few traces from the crashed threads (this is the most current error found): 
 Crashed: Thread: SIGSEGV  0x0000000000000000
#00 pc 0x75cc5cccd0
#01 pc 0x75cc5cccc4
#02 pc 0x75cbbe4ec4
#03 pc 0x75cbbe4384
#04 pc 0x75c767bfb0
#05 pc 0x75c767beec
#06 pc 0x75ce507adc
#07 pc 0x75c7679d9c
#08 pc 0x75c7679c50
#09 pc 0x75c21195d0
#10 pc 0x75bcba7698
#11 pc 0x75bc05ccb0
#12 pc 0x75bdb88fb4
#13 pc 0x75bdb88b00
#14 pc 0x75cfcbce24
#15 pc 0x75cfcbfd08
#16 pc 0x75cfcbe6a0
#17 pc 0x75cfcbe368
#18 pc 0x75cfcbe1d0
#19 pc 0x75ddc0c050
#20 pc 0x75e1bc1554
#21 pc 0xc67a0 libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#22 pc 0x25ea74 libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#23 pc 0x277b84 libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#24 pc 0x92038 libc.so (BuildId: f216ea105f08ade8b6c44c330ceacd65)
#25 pc 0x23968 libc.so (BuildId: f216ea105f08ade8b6c44c330ceacd65)

I also something find these ones but less frequent: 
null pointer dereference: SIGSEGV  0x0000000000000008
#00 pc 0x71b08ef51c
#01 pc 0x71b086af2c
#02 pc 0x50fac libc.so (BuildId: 83a645445a79ba5ba846eb0bc0e9cbef)

We also have a lot of SIGABRT like these: 
          Crashed: Thread: SIGABRT  0x0000000000000000
#00 pc 0x1d808 libc.so (BuildId: 6859390b71ada533f45843f0d88db865)
#01 pc 0x1d7dc libc.so (BuildId: 6859390b71ada533f45843f0d88db865)
#02 pc 0x1f360 libmono-android.release.so (BuildId: a32ab74a7903ac29677eaba52516f0df5457309d)
#03 pc 0x18b48 libmono-android.release.so (BuildId: a32ab74a7903ac29677eaba52516f0df5457309d)
#04 pc 0x1fec30 libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#05 pc 0x1fc040 libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#06 pc 0x1fba1c libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#07 pc 0x200da4 libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#08 pc 0x23d630 libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#09 pc 0x23baa4 libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#10 pc 0xf482c libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#11 pc 0xda878 libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#12 pc 0xbb2f8 libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#13 pc 0xbd934 libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#14 pc 0xc264c libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#15 pc 0xc1a6c libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#16 pc 0x152278 libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#17 pc 0x152f70 libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#18 pc 0x7780338500
#19 pc 0x777f6cf2e4
#20 pc 0x777faf2bf0
#21 pc 0x777f6ecbd0
#22 pc 0x4fffb0 libart.so (BuildId: 9e64a2227fb1f011124fdfacf025b9b0)
#23 pc 0x76e690ea70
#24 pc 0x76e690eed8
#25 pc 0x9a3f2c boot-framework.oat (BuildId: 59d24e79c600a981d8fe35084ebaaa6569cbb04e)
#26 pc 0xf1ba7c boot-framework.oat (BuildId: 59d24e79c600a981d8fe35084ebaaa6569cbb04e)
#27 pc 0x9a3e94 boot-framework.oat (BuildId: 59d24e79c600a981d8fe35084ebaaa6569cbb04e)
#28 pc 0x99f9d4 boot-framework.oat (BuildId: 59d24e79c600a981d8fe35084ebaaa6569cbb04e)
#29 pc 0x99bc04 boot-framework.oat (BuildId: 59d24e79c600a981d8fe35084ebaaa6569cbb04e)
#30 pc 0x62891c boot-framework.oat (BuildId: 59d24e79c600a981d8fe35084ebaaa6569cbb04e)
#31 pc 0x54c6cf libart.so (BuildId: 9e64a2227fb1f011124fdfacf025b9b0)
#32 pc 0x54c6d7 libart.so (BuildId: 9e64a2227fb1f011124fdfacf025b9b0)
#33 pc 0x526d88 libart.so (BuildId: 9e64a2227fb1f011124fdfacf025b9b0)
#34 pc 0xb7f94 libart.so (BuildId: 9e64a2227fb1f011124fdfacf025b9b0)
#35 pc 0x54c6cf libart.so (BuildId: 9e64a2227fb1f011124fdfacf025b9b0)
#36 pc 0x54c6d7 libart.so (BuildId: 9e64a2227fb1f011124fdfacf025b9b0)
#37 pc 0x1f74cc boot.oat (BuildId: beed0f987a4f8a1059e369c7fe30c8cc45ba162d)
#38 pc 0x44a9c8 libart.so (BuildId: 9e64a2227fb1f011124fdfacf025b9b0)
#39 pc 0xb7f94 libart.so (BuildId: 9e64a2227fb1f011124fdfacf025b9b0)
#40 pc 0x2db178 libart.so (BuildId: 9e64a2227fb1f011124fdfacf025b9b0)
#41 pc 0x47d9f4 libart.so (BuildId: 9e64a2227fb1f011124fdfacf025b9b0)
#42 pc 0x54c6d7 libart.so (BuildId: 9e64a2227fb1f011124fdfacf025b9b0)
#43 pc 0x54c6cf libart.so (BuildId: 9e64a2227fb1f011124fdfacf025b9b0)
#44 pc 0x44be5c libart.so (BuildId: 9e64a2227fb1f011124fdfacf025b9b0)
#45 pc 0x54c69e libart.so (BuildId: 9e64a2227fb1f011124fdfacf025b9b0)
#46 pc 0x60fc7c libart.so (BuildId: 9e64a2227fb1f011124fdfacf025b9b0)
#47 pc 0x324650 libart.so (BuildId: 9e64a2227fb1f011124fdfacf025b9b0)
#48 pc 0x87cd4 libandroid_runtime.so (BuildId: 986ad4e5d5a7f69cbb5317b081e27dcd)
#49 pc 0xeb34c libandroid_runtime.so (BuildId: 986ad4e5d5a7f69cbb5317b081e27dcd)
#50 pc 0x36f20 libbinder.so (BuildId: 9b088c8cd3c4dfe662f1cd0d350b9104)
#51 pc 0x41004 libbinder.so (BuildId: 9b088c8cd3c4dfe662f1cd0d350b9104)
#52 pc 0x41e04 libbinder.so (BuildId: 9b088c8cd3c4dfe662f1cd0d350b9104)
#53 pc 0x10e74 libutils.so (BuildId: 7ed1b99220236fd7277ce2df54565c33)
#54 pc 0x40d44 libbinder.so (BuildId: 9b088c8cd3c4dfe662f1cd0d350b9104)
#55 pc 0x41394 libbinder.so (BuildId: 9b088c8cd3c4dfe662f1cd0d350b9104)
#56 pc 0x40800 libbinder.so (BuildId: 9b088c8cd3c4dfe662f1cd0d350b9104)
#57 pc 0x6328c libbinder.so (BuildId: 9b088c8cd3c4dfe662f1cd0d350b9104)
#58 pc 0x11478 libutils.so (BuildId: 7ed1b99220236fd7277ce2df54565c33)
#59 pc 0x1e4afc libandroid_runtime.so (BuildId: 986ad4e5d5a7f69cbb5317b081e27dcd)
#60 pc 0x1e4afc libandroid_runtime.so (BuildId: 986ad4e5d5a7f69cbb5317b081e27dcd)
#61 pc 0x1135c libutils.so (BuildId: 7ed1b99220236fd7277ce2df54565c33)
#62 pc 0x78274 libandroid_runtime.so (BuildId: 986ad4e5d5a7f69cbb5317b081e27dcd)
#63 pc 0x10e74 libutils.so (BuildId: 7ed1b99220236fd7277ce2df54565c33)
        

Finally, we do seem to be seeing new ANRs than before our migration to .NET8. Some of them look like this: 
          m.ourcompany.ourapp (native):tid=3134 systid=3134 
#00 pc 0x4bf9c libc.so (syscall + 28) (BuildId: 83a645445a79ba5ba846eb0bc0e9cbef)
#01 pc 0x50528 libc.so (__futex_wait_ex(void volatile*, bool, int, bool, timespec const*) + 144) (BuildId: 83a645445a79ba5ba846eb0bc0e9cbef)
#02 pc 0x5d234 libc.so (sem_wait + 112) (BuildId: 83a645445a79ba5ba846eb0bc0e9cbef)
#03 pc 0x1eea08 libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#04 pc 0x1f4a08 libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#05 pc 0x277394 libmonosgen-2.0.so (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#06 pc 0x2773c8 libmonosgen-2.0.so (mono_threads_attach_coop + 24) (BuildId: b429a6405c79e0852cf3694861f0a6c5234a833a)
#07 pc 0x4adc
       
Let me know if you need anything else.
Thanks for your help.

**Comment by feric-isc** [comment]
@BrzVlad I believe this is a more complete Stacktrace using Crashlytics: 

[stacktrace_sigsegv.txt](https://github.com/user-attachments/files/20710345/stacktrace_sigsegv.txt)

Let me know if this can diagnose or if I can do anything else.
Thanks
PS: This is using 8.0.16 Runtime.

**Comment by BrzVlad** [comment]
@grendello This looks like a crash, potentially in the runtime. How do you recommend to symbolicate it ? Should we use `ndk-stack` ? We would need a 8.0.16 android arm64 image with debug symbols. Where to download it ?

**Comment by grendello** [comment]
@BrzVlad I usually do it by hand using `llvm-addr2line` but yes, `ndk-stack` should work fine. You can get the symbols by extracting `libmonosgen-2.0.so` from the apk, then using the `dotnet-symbol` tool (`dotnet-symbol libmonosgen-2.0.so`). This will place the debug symbols next to the code .so and `ndk-stack` should be able to symbolicate the trace.

**Comment by BrzVlad** [comment]
Thanks! @feric-isc Would it be possible to give these instructions a try, to see if you can symbolicate the crash report ?

**Comment by feric-isc** [comment]
@BrzVlad @grendello Thank you for the info. I'll talk with the team and get back to you.

**Comment by feric-isc** [comment]
@BrzVlad I tried using ndk-stack but had no success. I'm not sure why but I got an empty output even though I was able to generate the libmonosgen-2.0.so.dbg file...and I seemed to follow the instructions provided

This being said, I was able to do it using llvm-addr2line and replacing the addresses (using copilot). Here is what I got: [symbolicated_stacktrace.txt](https://github.com/user-attachments/files/20731386/symbolicated_stacktrace.txt)

Let me know if this is what you need or not. 
Thanks again

**Comment by BrzVlad** [comment]
Unfortunately the symbols are incorrect since the stack traces are impossible :(

**Comment by feric-isc** [comment]
@BrzVlad Damn AI :) Ok I'll do it manually. I'll get back to you early next week. Thanks

**Comment by BrzVlad** [comment]
There is a good chance the debug information is incorrect for whatever reason. Hopefully not

**Comment by grendello** [comment]
@feric-isc I wonder if name of the debug file was correct for your `libmonosgen-2.0.so`. There's a section in the ELF file, named `.gnu_debuglink`, which contains the expected name of the debug data file. You can check what it is by running `llvm-readelf --string-dump=.gnu_debuglink libmonosgen-2.0.so`

Also, `ndk-stack` might not have been able to recognize the stack trace format, it might be expecting the Android system crash report format. However, your stack traces are in a format that could be easily read by a shell script and then
fed to `llvm-addr2line`. Use this invocation of the utility: `llvm-addr2line -paCif -e libmonosgen-2.0.so <ADDRESS>`

**Comment by feric-isc** [comment]
@BrzVlad @grendello Ok I think I was able to get it. Here is the symbolized stack trace. Please let me know if this is what you're looking for: 

[stacktrace_symbolized.txt](https://github.com/user-attachments/files/20763285/stacktrace_symbolized.txt)

@grendello for your suggestion, I ran the llvm-readelf command. Here is the output: 
`>llvm-readelf --string-dump=.gnu_debuglink libmonosgen-2.0.so

String dump of section '.gnu_debuglink':
[     0] libmonosgen-2.0.so.dbg
[    18] ..\.`

**Comment by grendello** [comment]
@feric-isc I didn't verify the source code locations in your latest stack trace, but at the glance they look plausible :) @BrzVlad should be able to work with that :)

**Comment by feric-isc** [comment]
@BrzVlad Let me know if the traces are good for you. I can generate as many as you need. We have a ton every day. Thank you!
PS: We also have many SIGQUIT and SIGABRT...I don't know if they are related but the most common are the SIGSEGV.

**Comment by BrzVlad** [comment]
The stacktrace still doesn't look right to me. For example:
```
Thread:
#00 pc 0x4bf9c libc.so (BuildId: 83a645445a79ba5ba846eb0bc0e9cbef)
#01 pc 0x50528 libc.so (BuildId: 83a645445a79ba5ba846eb0bc0e9cbef)
#02 pc 0xb0980 libc.so (BuildId: 83a645445a79ba5ba846eb0bc0e9cbef)
#03 pc 0x13f8c libSystem.Native.so (BuildId: 6324e7c88ec2dab5f0850bcaf12686baf69d16a3)
#04 pc 0x7ee17190c8
#05 pc 0x7ee1718b98
#06 pc 0x7ee1718a9c
#07 pc 0x7ee17189a8
#08 pc 0x7ee177e8e4
#09 pc 0x7ee177bebc
#10 pc 0x7ee177b6f8
#11 pc 0x7ee177b080
#12 pc 0x7ee177ad08
#13 pc 0x7ee1779f94
#14 pc 0x7ee1779544
#15 pc 0x7ee177dd9c
#16 pc 0x7edd48b51c
#17 pc 0x7ee17cf40c
#18 pc 0x7ee17cfa24
#19 pc 0xcf7fc libmonosgen-2.0.so (BuildId: eef12aaa9c8ff06e938634117afc340e13dc429d)
    ==> mono_method_to_ir
    ==> /__w/1/s/src/mono/mono/mini/method-to-ir.c:11199:5
#20 pc 0x24a81c libmonosgen-2.0.so (BuildId: eef12aaa9c8ff06e938634117afc340e13dc429d)
    ==> mono_metadata_decode_row_col_raw
    ==> /__w/1/s/src/mono/mono/metadata/metadata.c:1479:23
    ==> mono_metadata_decode_row_col
    ==> /__w/1/s/src/mono/mono/metadata/metadata.c:1451:10
    ==> mono_metadata_localscope_from_methoddef
    ==> /__w/1/s/src/mono/mono/metadata/metadata.c:5219:30
#21 pc 0x26250c libmonosgen-2.0.so (BuildId: eef12aaa9c8ff06e938634117afc340e13dc429d)
    ==> mono_array_new_full_checked
    ==> /__w/1/s/src/mono/mono/metadata/object.c:5945:6
#22 pc 0xb1590 libc.so (BuildId: 83a645445a79ba5ba846eb0bc0e9cbef)
#23 pc 0x50fac libc.so (BuildId: 83a645445a79ba5ba846eb0bc0e9cbef)
```
This doesn't make sense to me and I don't understand why are there multiple resolved symbols for a single frame

**Comment by grendello** [comment]
@BrzVlad multiple symbols per frame might be caused by inlining, since debug information keeps data about inlined functions.

**Comment by BrzVlad** [comment]
Still, in this example `mono_metadata_decode_row_col_raw` cannot possibly call `mono_method_to_ir` not directly or ever. Also `mono_array_new_full_checked` ending up calling `mono_metadata_localscope_from_methoddef` seems implausible. Unless half the runtime code is inlined.. And these are the only symbols in the stacktrace for this thread for example ..

**Comment by grendello** [comment]
@BrzVlad it's possible the debug data is mismatched, somehow. However, the way to read the multiple entries is that the first one in a frame is the one called by the previous frame, so `mono_array_new_full_checked` calls `mono_metadata_decode_row_col_raw` which, eventually, calls `mono_method_to_ir` (via, possibly inlined, `mono_metadata_localscope_from_methoddef`). It really doesn't make much sense, just judging by the method names without looking at the source code though... :)

Another possibility is that Mono runtime was compiled without frame pointers and that the stack trace generator just gets lost.

**Comment by feric-isc** [comment]
Hello @grendello and @BrzVlad . Any progress? Anything I can do to help?
Thanks

**Comment by ahoffmandevelopment** [comment]
Also checking in to see if there's anything I can do to help. This is killing our Google Play metrics for a number of apps that we have.

**Comment by BrzVlad** [comment]
@ahoffmandevelopment  We would need a symbolicated stacktrace first to be able to at least understand what type of failure this is. Maybe you would have more success with this ?

> I usually do it by hand using llvm-addr2line but yes, ndk-stack should work fine. You can get the symbols by extracting libmonosgen-2.0.so from the apk, then using the dotnet-symbol tool (dotnet-symbol libmonosgen-2.0.so). This will place the debug symbols next to the code .so and ndk-stack should be able to symbolicate the trace.

**Comment by feric-isc** [comment]
@BrzVlad Are you able to confirm if you can or cannot work with my symbolicated traces? There seemed to be ambiguity between your comment and those of others? If not I can try again

**Comment by BrzVlad** [comment]
Sorry if I was ambiguous. Unfortunately the stack traces are not really useful since the resolved symbols are incorrect.

**Comment by feric-isc** [comment]
Hello @BrzVlad . Here are 2 additional symbolized stacktraces (with their non-symbolicated files they were generated from). 

[sigsegv1.txt](https://github.com/user-attachments/files/21221750/sigsegv1.txt)
[sigsegv1_symbolized.txt](https://github.com/user-attachments/files/21221751/sigsegv1_symbolized.txt)
[sigsegv2.txt](https://github.com/user-attachments/files/21221752/sigsegv2.txt)
[sigsegv2_symbolized.txt](https://github.com/user-attachments/files/21221753/sigsegv2_symbolized.txt)

Here is what I did to symbolicate. Let me know if something is not ok.

1. Downloaded stacktrace from Crashlytics Firebase (made sure to validate the crash was with our latest version that was built with .NET SDK 8.0.410
2. Unzipped libmonosgen-2.0.so
3. Ran command dotnet-symbol libmonosgen-2.0.so. Generated dotnet-symbol libmonosgen-2.0.dbg
4. Ran the following python script to generate the symbolicated traces

```python
import re
import subprocess

# === Configuration ===
STACKTRACE_FILE = "sigsegv2.txt"
OUTPUT_FILE = "sigsegv2_symbolized.txt"
DBG_FILE = r"libmonosgen-2.0.so.dbg"
LLVM_SYMBOLIZER = "llvm-symbolizer.exe" 

# === Regex to match libmonosgen lines ===
libmono_re = re.compile(r'#\d+\s+pc\s+(0x[0-9a-f]+)\s+libmonosgen-2\.0\.so', re.IGNORECASE)

def symbolize_address(addr: str) -> str:
    """Call llvm-symbolizer with one address and return the symbolized result."""
    try:
        result = subprocess.run(
            [LLVM_SYMBOLIZER, f"--obj={DBG_FILE}", addr],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            check=True
        )
        return result.stdout.strip()
    except subprocess.CalledProcessError as e:
        return f"[symbolization failed: {e.stderr.strip()}]"

def main():
    with open(STACKTRACE_FILE, "r", encoding="utf-8") as f_in, open(OUTPUT_FILE, "w", encoding="utf-8") as f_out:
        for line in f_in:
            f_out.write(line)
            match = libmono_re.search(line)
            if match:
                addr = match.group(1)
                symbolized = symbolize_address(addr)
                for sym_line in symbolized.splitlines():
                    f_out.write(f"    ==> {sym_line}\n")
    print(f"✅ Symbolized output written to: {OUTPUT_FILE}")

if __name__ == "__main__":
    main()
```

**Comment by feric-isc** [comment]
Hi @BrzVlad. Are these traces better?

**Comment by BrzVlad** [comment]
The stacktraces are definitely correct this time, thanks for taking the time to symbolicate the reports. There are no symbols for `libc.so` and `libart.so` so I can't be 100% that there are no other issues somewhere else outside of the runtime. I've seen crashes before in the art GC or in other places. At least we know there no crashes in the .NET runtime.

Having said that, if we are to assume that the first reported thread is the one that is crashing (I think this should be the case but I don't have full trust in these reports), this points to a null pointer dereference in managed code crashing the application. This normally should be impossible, a NRE should be thrown, unless the signal handlers have been tampered with. We've recently uncovered such an issue, fixed by https://github.com/dotnet/runtime/pull/118085. So if your application uses `Process.Start` or `System.Net.NetworkInformation.Ping`, this means you are hitting the above mentioned issue.

**Comment by feric-isc** [comment]
Thanks for your feedback @BrzVlad . We checked and we don't seem to use Process.Start or System.Net.NetworkInformation.Ping. Is that something you deduced from our stacktraces? If we don't use them, do you still think it could be related to the issue you mentioned? I seem to see that it has been backported in .NET9. So if we update to latest of .NET9 and we still have the issue, what then? We are planning to do so in the next month or so...

**Comment by BrzVlad** [comment]
The version containing the mentioned fix will be 9.0.10. So it will take a while before you can try it out (I think around 1.5 months). To me it just feels like the only plausible explanation, but I wouldn't be overly confident about it. In order to validate this in the meantime, you could try to use the following environment flag: `MONO_DEBUG=explicit-null-checks` (this avoids usage of signals when throwing NRE). You should be able to tweak the environment by adding the value to a new file and then include this file in the build as an android environment file. For example:
```
<ItemGroup>
        <AndroidEnvironment Include="Platforms\Android\Environment.txt" />
</ItemGroup>
````

You might want to check at runtime with `Environment.GetEnvironmentVariable` to see that the env var is set correctly, there have been some issues in the past (only with debug builds I think).

**Comment by bnSonic** [comment]
I cannot say too much - but I see these crashes in the google play console (it raises the crashes to above 2% so I'm getting warnings from google)

```
[split_config.arm64_v8a.apk!libmonosgen-2.0.so]
SIGSEGV
```

currently I have no idea if there is more information available on the google console site …

I'm using the .net9 / maui 9.0.120 versions

**Comment by slepmog** [comment]
Also plagued by this. As soon as we were forced to port from Xamarin to .net9.0-android, these started.

```none
11:13:16.264 19622 19622 F DEBUG   : signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x0000000000000003
11:13:16.264 19622 19622 F DEBUG   : Cause: null pointer dereference
11:13:16.264 19622 19622 F DEBUG   :     x0  0000000000000000  x1  0000000000000093  x2  00000070c001a60c  x3  00000070c001a548
11:13:16.264 19622 19622 F DEBUG   :     x4  0000000000000000  x5  b4000070b33c8be2  x6  0000000000000029  x7  0000000000000029
11:13:16.264 19622 19622 F DEBUG   :     x8  b400007088418000  x9  0000000000000000  x10 b4000070843caadd  x11 0000000000000000
11:13:16.264 19622 19622 F DEBUG   :     x12 0000000000000098  x13 0000000000000084  x14 0000000000000028  x15 00000071e366f792
11:13:16.264 19622 19622 F DEBUG   :     x16 00000070c67a3ec8  x17 00000070c66ff870  x18 00000070bf8ac000  x19 b4000070b33c8b80
11:13:16.264 19622 19622 F DEBUG   :     x20 0000000000000093  x21 b4000070883dc048  x22 0000000000000026  x23 00000070c64e0d73
11:13:16.264 19622 19622 F DEBUG   :     x24 00000071e374ecf0  x25 00000070c64dba73  x26 b40000706d41db50  x27 0000000000000013
11:13:16.264 19622 19622 F DEBUG   :     x28 0000000000000178  x29 00000070c001a620
11:13:16.264 19622 19622 F DEBUG   :     lr  00000070c67050e0  sp  00000070c001a5d0  pc  00000070c66ff87c  pst 0000000020001000
11:13:16.264 19622 19622 F DEBUG   : backtrace:
11:13:16.264 19622 19622 F DEBUG   :       #00 pc 000000000025087c  /data/app/~~9OHsKnAIfChr21VlFWaEzQ==/com.redacted.appname-djIIp-UazvdZ9PDIjVMj3g==/lib/arm64/libmonosgen-2.0.so (mono_guid_to_string_minimal+12) (BuildId: 74bd9f1f34dbeeaa747197639b575cbcdecba5be)
11:13:16.264 19622 19622 F DEBUG   :       #01 pc 00000000002560dc  /data/app/~~9OHsKnAIfChr21VlFWaEzQ==/com.redacted.appname-djIIp-UazvdZ9PDIjVMj3g==/lib/arm64/libmonosgen-2.0.so (mono_debug_print_stack_frame+292) (BuildId: 74bd9f1f34dbeeaa747197639b575cbcdecba5be)
11:13:16.264 19622 19622 F DEBUG   :       #02 pc 000000000026eca8  /data/app/~~9OHsKnAIfChr21VlFWaEzQ==/com.redacted.appname-djIIp-UazvdZ9PDIjVMj3g==/lib/arm64/libmonosgen-2.0.so (BuildId: 74bd9f1f34dbeeaa747197639b575cbcdecba5be)
11:13:16.264 19622 19622 F DEBUG   :       #03 pc 0000000000293498  /data/app/~~9OHsKnAIfChr21VlFWaEzQ==/com.redacted.appname-djIIp-UazvdZ9PDIjVMj3g==/lib/arm64/libmonosgen-2.0.so (BuildId: 74bd9f1f34dbeeaa747197639b575cbcdecba5be)
11:13:16.264 19622 19622 F DEBUG   :       #04 pc 0000000000270508  /data/app/~~9OHsKnAIfChr21VlFWaEzQ==/com.redacted.appname-djIIp-UazvdZ9PDIjVMj3g==/lib/arm64/libmonosgen-2.0.so (BuildId: 74bd9f1f34dbeeaa747197639b575cbcdecba5be)
11:13:16.264 19622 19622 F DEBUG   :       #05 pc 00000000000f5108  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+208) (BuildId: bddac2439f5f0b431ea53a2f053677d9)
11:13:16.264 19622 19622 F DEBUG   :       #06 pc 000000000008eafc  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+68) (BuildId: bddac2439f5f0b431ea53a2f053677d9)
```

The entire warehouse operation is affected by this, and we cannot go back to Xamarin because that no longer even builds, because of the minimum API restriction.

**Comment by BrzVlad** [comment]
@slepmog It's not clear if your crash is related. Could you upgrade to latest .net 10 first ?

**Comment by slepmog** [comment]
@BrzVlad I thought it was related because it's the same type of crash (signal 11 (SIGSEGV), code 1 (SEGV_MAPERR)) in the same lib (libmonosgen-2.0.so) with the same no way to catch it (the process is killed by the OS instead of triggering our `try`/`catch`).

And we would need to switch to VS 2026 to try .net 10.  
In the past 10 years our experience has been that with .NET Android development, when you try to work with an existing project in a later version of Visual Studio, it breaks in unfathomable ways and fails to compile with hundreds of errors that don't make sense, so switching the VS version has always been a full-time, one-way, porting project.
